src/uct/tcp/tcp_ep.c

```diff
@@ -660,7 +689,49 @@ uct_tcp_ep_comp_recv_am(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
     uct_iface_invoke_am(&iface->super, hdr->am_id, hdr + 1, hdr->length, 0);
 }
 
-unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
+/* Forward declaration - the function depends on AM send
+ * functions implemented below */
+static void uct_tcp_ep_post_put_ack(uct_tcp_ep_t *ep,
+                                    uct_tcp_ep_put_req_t *put_req);
+
+static inline void uct_tcp_ep_handle_put_req(uct_tcp_ep_t *ep,
+                                             uct_tcp_ep_put_req_t *put_req,
+                                             size_t extra_recvd_length)
+{
+    size_t copied_length;
+
+    if (!put_req->length) {
+        uct_tcp_ep_post_put_ack(ep, put_req);
+        return;
+    }
+
+    ucs_assert(put_req->addr != 0);
+
+    copied_length    = ucs_min(put_req->length, extra_recvd_length);
+    memcpy((void*)(uintptr_t)put_req->addr,
+           UCS_PTR_BYTE_OFFSET(ep->rx.buf, ep->rx.offset),
+           copied_length);
+    ep->rx.offset   += copied_length;
+    put_req->addr   += copied_length;
+    put_req->length -= copied_length;
+
+    if (!put_req->length) {
+        uct_tcp_ep_post_put_ack(ep, put_req);
+        return;
+    }
+
+    ucs_assert(ep->rx.offset == ep->rx.length);
+    uct_tcp_ep_ctx_rewind(&ep->rx);
+    memcpy(ep->rx.buf, put_req, sizeof(*put_req));
+    ep->ctx_caps |= UCS_BIT(UCT_TCP_EP_CTX_TYPE_PUT_RX);
+}
+
+static inline void uct_tcp_ep_handle_put_ack(uct_tcp_ep_t *ep)
```

Comment:
no need for this func, just invoke ` uct_tcp_ep_outstanding_dec(ep)` in place

Reply:
done